feat: allow passing readCache as an object

garcianavalon opened the pull request:
*This is still a Work In Progress, do not merge :)*

This PR allows passing a `readCache object` to `getPostGraphileBuilder`. The functionality is backwards compatible, passing a `readCache` string with a path to a file will still work on the same way.

garcianavalon commented:
@benjie please take a look when you have time to see if the functionality is what you required. If thats ok, I'll write the tests

benjie commented:
Looking good 👍

benjie commented on packages/postgraphile-core/src/index.ts:
```diff
@@ -281,6 +281,8 @@ const getPostGraphileBuilder = async (
       );
     }
+  } else if (readCache && typeof readCache === "object") {
+    memoizeCache = readCache;
```
```suggestion
    memoizeCache = readCache;
  } else {
    throw new Error(`'readCache' not understood; expected string or object, but received '${typeof readCache}'`);
```

garcianavalon commented:
Thanks for the suggestion, added it.

I expect to finish tests by this week

benjie commented:
Great :+1:

garcianavalon commented:
Hi @benjie during the weekend I'll get back to this now that the test PR is merged. 

At some point, travis started failing. Checking at the logs, looks like its a linter error. I haven't seen any specifics about linting on the CONTRIBUTING docs, I assumed it would be some configuration stored in the repo. Is there anything else I need to set up on my environment regarding linters?

benjie commented:
@garcianavalon Good spot; we should add a `yarn prettier:fix` command to root. You can run prettier against the relevant file via

```
npx prettier --write packages/postgraphile-core/src/index.ts
```

My editors (Vim and VSCode) automatically fix any lint errors on save so I don't think about it. Please feel free to open a dedicated issue about this :+1:

garcianavalon commented:
@benjie I think this PR is ready (removing WIP)

- Fixed an error that caused getPostGraphileBuilder to throw an error if cache was null or undefined
- Pulled master for the readCache tests from the other PR. All tests run successfully
- Added new tests to verify object input and error throwing with invalid input
- Refactored test into describe blocks for readability (I had to update the snapshots due to the change in test name, in case you wonder why they appear as modified. I made sure to run tests *before* changing them though)

I'm waiting for good ol' travis to finish running the pipeline, but I think the linter errors have gone away on their own. I also use VSCode and is also linting on save automagically, an error might have sliped in when adding suggestions I guess.

benjie commented on packages/postgraphile-core/__tests__/unit/read-cache.test.js:
```diff
@@ -67,22 +45,109 @@ test("when no readCache flag, persistentMemoizeWithKey should be undefined", asy
+    } catch (e) {
+      error = e;
+    }
+    expect(error).toMatchSnapshot();
```
This isn't actually throwing - the error is `undefined`. You might want to add `expect(error).toBeTruthy()` to each of these tests, snapshots should only be used for looking for changes over time (e.g. error message text, new error properties, etc) rather than asserting behaviour.

garcianavalon replied:
you are right! I'll add that check

garcianavalon replied:
done

benjie commented on packages/postgraphile-core/src/index.ts:
```diff
@@ -265,22 +265,35 @@ export const getPostGraphileBuilder = async (
+          `Failed to parse cache file '${readCache}', perhaps it is corrupted? ${e}`
+        );
+      }
+    } else if (typeof readCache === "object") {
```
```suggestion
    } else if (typeof readCache === "object" && !Array.isArray(readCache)) {
```

garcianavalon replied:
I always forget how stupid js can be, thanks for the addition

garcianavalon commented:
> LGTM; I'm a little concerned that the cache might be modified but so long as it's not combined with watch mode that should be fine (and they shouldn't be combined... that doesn't make sense!). Tiny tweak to logic and tests required.

we could deep clone the object instead of direct assignment if you think thats a concern

benjie commented:
I considered it, but 🤷‍♂️ I think wait for someone to complain ;) Performance-wise avoiding a deep copy is preferable.

benjie commented:
The snapshots need updating (run the tests with `-u` to update the snapshots):

```
  Snapshot name: `when readCache is not String or Object, getPostGraphileBuilder should error when its Array 1`
    - Snapshot
    + Received
    - undefined
    + [Error: 'readCache' not understood; expected string or object, but received 'object']
      137 |       error = e;
      138 |     }
    > 139 |     expect(error).toMatchSnapshot();
          |                   ^
      140 |   });
      141 |   test("when its Number", async () => {
      142 |     // call our method with invalid readCache value and check error
      at Object.toMatchSnapshot (__tests__/unit/read-cache.test.js:139:19)
```

garcianavalon commented:
sorry @benjie I got a job spike and I've been quite busy. I expect to finish it this week

benjie commented:
No worries :+1:

garcianavalon commented:
@benjie updated snapshots, added assert to verify that errors are not  `undefined`

benjie commented on packages/postgraphile-core/src/index.ts:
```diff
@@ -265,22 +265,35 @@ export const getPostGraphileBuilder = async (
+    } else {
       throw new Error(
-        `Failed to parse cache file '${readCache}', perhaps it is corrupted? ${e}`
+        `'readCache' not understood; expected string or object, but received '${typeof readCache}'`
```
```suggestion
        `'readCache' not understood; expected string or object, but received '${Array.isArray(readCache) ? 'array' : typeof readCache}'`
```

garcianavalon commented:
Applied your suggestion, updated snapshots 👍

benjie commented:
Great; thanks!

CI is failing because one of the files doesn't pass the "prettier" check. It seems we don't have a `lint:fix` command :flushed: Can you run something like:

```
prettier --write packages/postgraphile-core/src/index.ts
```

please? Sorry :smile:

I should really make this easier for people to comply with.

garcianavalon commented:
Haha sure no problem!

On Mon, Sep 9, 2019, 20:54 Benjie Gillam <notifications@github.com> wrote:

> Great; thanks!
>
> CI is failing because one of the files doesn't pass the "prettier" check.
> It seems we don't have a lint:fix command 😳 Can you run something like:
>
> prettier --write packages/postgraphile-core/src/index.ts
>
>
> please? Sorry 😄
>
> I should really make this easier for people to comply with.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/graphile/graphile-engine/pull/479?email_source=notifications&email_token=AAK6Y3POQ4Y7PIWKDLJVJP3QI2LX7A5CNFSM4H3R3WEKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6IVGPI#issuecomment-529617725>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAK6Y3L2IPCHUBMS4XUK6YDQI2LX7ANCNFSM4H3R3WEA>
> .
>

garcianavalon commented:
Done! lets see if travis likes it now :D

benjie commented:
:tada: CI passes :+1:

benjie commented:
Thanks @garcianavalon; merged! :tada:

garcianavalon commented:
💃 awesome!
